source/common/http/conn_pool_grid.h

```diff
@@ -103,12 +106,17 @@ class ConnectivityGrid : public ConnectionPool::Instance,
     // The decoder for the original newStream, needed to create streams on subsequent pools.
     Http::ResponseDecoder& decoder_;
     // The callbacks from the original caller, which must get onPoolFailure or
-    // onPoolReady unless there is call to cancel().
-    ConnectionPool::Callbacks& inner_callbacks_;
+    // onPoolReady unless there is call to cancel(). Will be nullptr if the caller
+    // has been notified while attempts are still pending.
+    ConnectionPool::Callbacks* inner_callbacks_;
     // The timer which tracks when new connections should be attempted.
     Event::TimerPtr next_attempt_timer_;
     // The iterator to the last pool which had a connection attempt.
     PoolIterator current_;
+    // True if the HTTP/3 attempt failed.
+    bool http3_attempt_failed_{};
```

Comment:
nit: it feels weird to have these named "http3" vs "tcp", should we instead say "quic" vs "tcp"?

Reply:
"http3" is definitely the right name here because the underlying pool is an "HTTP3" pool, not a QUIC pool (both the name of the class and the type of connections it creates). "tcp" is more debatable. The underlying class is HttpConnPoolImplMixed which we sometimes (I'm told) call "the ALPN pool". I originally used "alpn" instead of "tcp" but alyssa pointed out that HTTP/3 does ALPN so that's not terribly accurate either.

I think "tcp" is the least inaccurate name so I'm inclined to keep it, at least for now.

Reply:
I'm not sure that's right. I think of this as a QUIC pool which happens to negotiate HTTP/3 via ALPN. If we were to standardize HTTP/4 which also runs over QUIC (with disambiguation via ALPN like HTTP/1.1 vs HTTP/2) then it would use this pool, not a separate pool. Right?

Reply:
Hm.  I don't think that QUIC helps make things more readable, specifically because the pool does not vend QUIC streams (the Envoy class StreamEncoder), it vends HTTP/3 streams (the Envoy class RequestEncoder).  The name of the Pool's class is "Http3ConnPoolImpl", not "QuicConnPoolImpl". If at some point this pool vends some other connections than HTTP/3 (say, HTTP/4) we could certainly consider renaming it. But it seems sensible to keep the names consistent in both classes.

That being said, if we feel strongly and want to rename both of them (and perhaps other associated classes) that's definitely an option. Though I'd prefer we do so in a follow up.

Reply:
Well the TCP pool doesn't vend TCP streams, nor does it vend ALPN streams :P

I'm OK with punting the renaming to a follow-up though, so I'll stop complaining :)

Reply:
Definitely up for punting :)

> Well the TCP pool doesn't vend TCP streams, nor does it vend ALPN streams :P

Yes! I totally agree with you here! I think the thing which is poorly named here is "tcp" :) Naming is hard! (Maybe we need to start calling things Legacy HTTP and Modern HTTP :)